dask/array/core.py

```diff
@@ -4435,7 +4435,27 @@ def concatenate_axes(arrays, axes):
     return concatenate3(transposelist(arrays, axes, extradims=extradims))
 
 
-def to_hdf5(filename, *args, **kwargs):
+def _to_hdf5(filename, data, chunks, **kwargs):
+    """ Function to wrap context managed hdf5 save so it can be defered. """
+    import h5py
+
+    with h5py.File(filename, mode="a") as f:
+        dsets = [
+            f.require_dataset(
+                dp,
+                shape=x.shape,
+                dtype=x.dtype,
+                chunks=tuple([c[0] for c in x.chunks]) if chunks is True else chunks,
+                **kwargs,
+            )
+            for dp, x in data.items()
+        ]
+        store(list(data.values()), dsets, compute=True)
```

Comment:
Hmmm, this isn't quite correct - the `store` function calls `dask.compute` internally, and by delaying `_to_hdf5` you end up wrapping a `dask.compute` call in a `dask.compute` call. Having `compute` call `compute` results in poor performance, and isn't something we'd really want to do.

Thinking structurally about what we're trying to accomplish here - the simplest approach would be to have each task that writes a chunk to the hdf5 file opens the file, writes the chunk, then closes the file. We might accomplish this by creating a proxy object that on `__setitem__` opens the hdf5 file, writes to it, and closes it. I'm not sure if there would be a performance impact with that or not - if there is, we might make the proxy object "smarter" to cache opens on the same process, but we can deal with that later if the need arises.

Example proxy object (untested)

```python
class HDFDatasetProxy:
    def __init__(self, filename, **kwargs):
        self.filename = filename
        self.kwargs = kwargs

    def __setitem__(self, key, val):
        with h5py.File(filename, mode="a") as f:
            ds = f.require_datset(**self.kwargs)
            ds[key] = val
```

And the setup code above:

```python
dsets = [
    HDF5DatasetProxy(
        filename,
        name=dp,
        shape=x.shape,
        dtype=x.dtype,
        chunks=tuple([c[0] for c in x.chunks]) if chunks is True else chunks,
        **kwargs,
    )
    for dp, x in data.items()
]
store(list(data.values()), dsets, compute=compute)
```

Note that this would no longer be called from inside a delayed function, but would be back in `to_hdf` proper itself.

Does that make sense?

Reply:
Sure that makes sense! I think I'd confused myself around what the `store` was actually deferring and was worried about the store being split in to chunks somehow.